design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,317 @@
+# KubeVirt Feature Lifecycle
+
+## Summary
+
+KubeVirt requires a clear policy on how features are introduced,
+evaluated and finally graduated or removed.
+
+This proposal defines the steps and policies to follow in order
+to manage a feature and its lifecycle in KubeVirt.
+
+The proposal is focusing on introducing features in
+a stable API (CRD) version, e.g. `kubevirt.io/v1`.
+
+## Overview
+
+### Motivation
+KubeVirt has grown into a matured virtualization management solution
+with a large set of features.
+
+New features are being proposed and added regularly to its portfolio.
+
+With time, the challenge of supporting and maintaining such a large
+set of features raised the need to re-examine their relevance.
+It also raised the need to examine with more care features graduation.
+
+The KubeVirt community has tried to control the flow of features
+informally through feature-gates, similar to Kubernetes.
+However, as time passed, several challenges presented themselves:
+- Evaluated features rarely got graduated to GA or removed, causing
+  feature consumption to be risky for users and a maintenance burden
+  for the project contributors.
+  > **Note**: As of this writing (pre v1.2), there are 37 FGs,
+  > out of which 5 GA-ed and 2 marked for deprecation.
+  > For more information, explore the
+  > [source](https://github.com/kubevirt/kubevirt/blob/5ff12ae931cefd81514ec96f97a189ab2c179ad7/pkg/virt-config/feature-gates.go).
+- We do not have agreed-upon procedures on how and when features
+  graduate or discontinue.
+  This causes each feature to take different approaches, possibly
+  surprising users.
+
+We conclude that the current use of FGs is insufficient
+due to the lack of well-defined processes and policy on how a feature
+should progress in its lifecycle.
+
+Eventually we would like to see features being evaluated carefully
+before they are introduced, while they are experimented with and
+proven to be actually in use (and useful) before graduating.
+
+> **Note**: Once a feature graduates, it is included in a
+> General Availability (GA) release with its functionality available
+> to all users. GA features need to comply with [semver](https://semver.org/)
+> which add constraints on their ability to change (including deprecation).
+
+### Goals
+- Define the process a feature needs to pass in order to be
+  Generally Available.
+- Define the process a feature needs to pass in order to be removed.
+- Provide policies and rules on how to manage a feature during its
+  lifetime.
+
```

Comment:
What about existing features that are feature gated?

Reply:
is there a concern here that in order to get the current features adhere to this policy will be a non-trivial amount of work? for example getting feature configuration and feature gate separated for current FG. Should we track this in an open issue as part of this proposal?

Reply:
That is an implementation step that should follow the acceptance of this process.
I guess that if a problem is detected, the topic will be discussed in the context of a specific feature.

Does that sound reasonable?

Reply:
Sure, sounds good.

Reply:
I am fine as long as we consider this before any "implementation" of the policy.